Added property based tests

edescourtis opened the pull request:
- Check that we always return to passive mode (for both give_away and heir)
- Check that we reply 'full' when the PO Box has reached capacity
- Fixed incorrect documentation in README.md (missing Pid in message)

ferd commented on .travis.yml:
```diff
@@ -8,4 +8,9 @@ otp_release:
 script:
-  - rebar3 ct --readable=false
\ No newline at end of file
+  - rebar3 ct --cover && rebar3 proper -c -n 1000 && rebar3 cover
```
`rebar3 do ct -c, proper -c -n 1000, cover -v`

ferd commented on README.md:
```diff
@@ -66,7 +66,7 @@ buffer and drop them when necessary.
 The notify state is enabled by the user by calling the PO Box. Its sole
 task is to verify if there is any message in the buffer. If there is, it
-will respond to the PO Box's owner with a `{mail, new_data}` message
+will respond to the PO Box's owner with a `{mail, BoxPid, new_data}` message
```
Is this a new backwards incompatible change, or the doc was always wrong?

edescourtis replied:
Looks like that. I don't recall changing it. I will double check.

edescourtis replied:
Okay so I found when it was changed commit 9325f4abafccb5dba7e697bbf3856b8eccb22ac5 . Which was commited by you on 2013-08-14 09:31:17.

ferd replied:
Sounds good, so I just had not updated the doc/messed up versioning back then.

ferd commented on rebar.config:
```diff
@@ -5,3 +5,19 @@
+            %% hex
+            {proper, "1.3.0"},
+            %% newest from master
+            {proper, {git, "https://github.com/proper-testing/proper.git",
```
Drop this one, we'll only use the one from hex

edescourtis replied:
Done.

ferd commented on rebar.config:
```diff
@@ -5,3 +5,19 @@
 ]}.
+
+%% the plugin itself
+{plugins, [rebar3_proper]}.
```
use `project_plugins` so we won't fetch the plugins as a transitive dep

edescourtis replied:
ok

ferd reviewed (commented):
Please untrack the test/.rebar3/erlcinfo file. This is a build artifact of the plugin using an old compiler API.

ferd commented on test/prop_pobox.erl:
```diff
@@ -0,0 +1,111 @@
+        {ok, SCPid} = pobox_state_changer:start_link(self(), Behaviours, CrashVector, BufType),
+        Box = pobox_state_changer:get_pobox(SCPid),
+        %io:format("~p~n", [Scenario]),
+        Procs = [erlang:spawn_opt(node(), fun() -> catch do_messaging_behaviour(Box, MessagingBehaviour) end, [monitor])
```
you can just use `spawn_monitor(fun() -> ... end)` and generally wouldn't need to `catch` anything since you're looking ofr things going down.

edescourtis replied:
Alright, consider it done.

ferd commented on test/prop_pobox.erl:
```diff
@@ -0,0 +1,111 @@
+        case is_process_alive(SCPid) of
+            true -> erlang:exit(SCPid, kill);
+            false -> ok
+        end,
```
should we expect that process to be dead or not? Both cases are made acceptable but I'm not too sure why

edescourtis replied:
I would have to add a stop behaviour to all scenarios.  I will do that instead.

edescourtis replied:
Actually no that might not work. There is no way to know the behaviour of the messaging so we need some way to guarantee that the process terminates.

ferd replied:
Well the question then is: why not just always kill the process? What is it you're trying to check as a property?

edescourtis replied:
This is because I don't really care about that process. I just want to exercise the pobox.

edescourtis commented:
Found an interesting bug, looking for the cause. Would you mind giving me a hand fixing this one?:
```
=ERROR REPORT==== 4-Dec-2018::16:17:23 ===
** State machine <0.30392.0> terminating
** Last event = {{call,{<0.30397.0>,#Ref<0.0.2.24189>}},{resize,1}}
** When server state  = {passive,
                            {state,
                                {buf,keep_old,5,5,2,
                                    {[{msg,<<>>}],[{msg,<<>>}]}},
                                <0.3.0>,undefined,undefined,<0.3.0>,undefined,
                                undefined,undefined,<0.30392.0>}}
** Reason for termination = error:badarg
** Callback mode = state_functions
** Stacktrace =
**  [{queue,split,
            [4,{[{msg,<<>>}],[{msg,<<>>}]}],
            [{file,"queue.erl"},{line,325}]},
     {pobox,drop,4,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,499}]},
     {pobox,resize_buf,2,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,448}]},
     {pobox,handle_call,4,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,344}]},
     {gen_statem,call_state_function,5,[{file,"gen_statem.erl"},{line,1223}]},
     {gen_statem,loop_event,6,[{file,"gen_statem.erl"},{line,995}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]
```

See [https://travis-ci.com/edescourtis/pobox/jobs/162727189](https://travis-ci.com/edescourtis/pobox/jobs/162727189)

ferd commented:
So for the bug you got:

```
** Last event = {{call,{<0.30718.0>,#Ref<0.0.1.35305>}},{resize,1}}
** When server state  = {passive,
                            {state,
                                {buf,keep_old,5,5,2,
                                    {[{msg,<<>>}],[{msg,<<>>}]}},
                                <0.30712.0>,undefined,undefined,<0.30712.0>,
                                undefined,undefined,undefined,<0.30713.0>}}
[{queue,split,
            [4,{[{msg,<<>>}],[{msg,<<>>}]}],
            [{file,"queue.erl"},{line,325}]},
     {pobox,drop,4,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,499}]},
     {pobox,resize_buf,2,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,448}]},
     {pobox,handle_call,4,
            [{file,"/home/travis/build/edescourtis/pobox/_build/test/lib/pobox/src/pobox.erl"},
             {line,344}]},
     {gen_statem,call_state_function,5,[{file,"gen_statem.erl"},{line,1223}]},
     {gen_statem,loop_event,6,[{file,"gen_statem.erl"},{line,995}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]
```

The state record is equivalent to:

```erlang
#buf{type = keep_old, 
         max = 5,
         size = 5,
         drop = 2, ...} 
```

The failing call is, strictly speaking, `queue:split(4, QueueWith2Elements)` and the error is due to asking to split a queue to a size larger than it is. The code that calls that is here: https://github.com/ferd/pobox/blob/2930711854c40a70622794070e2b76c1808467f4/src/pobox.erl#L485-L500

So this points at an error in the handling of the size of the queue. The sequence of operations was:

```
{1,[{wait,0}],[
  {0,async,{msg,<<>>}},{0,async,{msg,<<>>}},
  {1,async,{resize,5}},
  {0,async,{msg,<<>>}},
  {1,async,{resize,1}},
  {0,async,{msg,<<>>}},{0,async,{msg,<<>>}},{0,async,{msg,<<>>}},{0,async,{msg,<<>>}}
 ],keep_old}
```
It's a bit odd to use only types for the generator since it won't allow specific constraits (`LET` or `SUCHTHAT`). Also what you're doing is essentially rolling up a stateful property by hand, rather than using one of the existing handlers. 

Looking at it, I have the following understanding:

1. what is a crash vector?
2. on transfer, `{wait, 0}` is the only specified behaviour which means that on any message it will wait 0 ms and then stop on the message after that?
3. you'll run through a list of operations (send 2 messages, resize to 5, send 1 message, resize to 1, send 4 messages)

So clearly we have a problem since the state mentioned we had 5 messages, but only 3 could have been sent until then. It also mentioned 2 messages dropped, whereas there would be no reason to drop any.

This points to the resizing as the likely bad operation, since otherwise the 5 can't come in from anywhere. You'll note that your initialization call specifies no `max` value (`{ok, Box} = pobox:start_link(#{size => 10, heir=>Heir, type => BufType}),`), which was not possible under the old API. The problem is in the defaults chosen:

```
default_opts() ->
  #pobox_opts{owner=self(), initial_state=notify, type=queue}.
```

Later on, the call is made to this: 

https://github.com/ferd/pobox/blob/2930711854c40a70622794070e2b76c1808467f4/src/pobox.erl#L437-L446

You'll notice that a comparison is made between the current size and the max size. However the atom `undefined` compares as greater than any number whatsoever.

My guess is that since you don't specify a max buffer value, your chosen default of `undefined` means that the existing code assumes that the new size is what the final buffer size will be, and therefore sets `size=NewMax` where `NewMax` is 5, despite the actual queue only having a size 2.

The bug, if I had to choose it, would be to allow an interface that does not enforce a minimal buffer size, something that is integral to how POBox works. I would also say that a default is not as good as asking the user to always choose a size. If it's not in there, initial validation should fail.

ferd commented:
Also that's a cool example of a property test catching a small but subtle bug. This can only get triggered in the case where you do two consecutive resizes, and the first resize goes to a greater size than the following one when accounting for all the data in the mailbox :)

ferd commented:
Ah wait, there was a size of 10 there in the argument I overlooked. I think you wanted it to be named `max` instead of `size`. In fact, `size` should not be configurable and should always just be 0 since it's a cache for the amount of entries in the state

edescourtis commented:
`size` is the max value here I can change it to max_size to match the documentation.

edescourtis commented:
All we do with the `size` is call `buf_new(Type, Size)` in `init/1`.  On top of that dumping the `state` reveals the `buf` record to be in the following state: `{buf,keep_old,10,0,0,{[],[]}}`. That part looks correct to me.

ferd commented:
Ah, I had skipped that `init` step that adds one layer of indirection. The resize functionality sounds like the most likely culprit, but 

https://github.com/ferd/pobox/blob/2930711854c40a70622794070e2b76c1808467f4/src/pobox.erl#L436-L446

This shouldn't be a problem given I assume the state of the buffer should be `{buf,keep_old,10,2,0, QLen2}` by the time we call the resize. To get to `{buf,keep_old,5,5,2,QLen2}` we have no choice but to corrupt the data since the size won't match the queue length.

I haven't run the thing, but I am really suspecting a corrupting action somewhere. At this point I'd recommend finding the repeatable case (see if `rebar3 proper --retry` always fail) and then try to run it by calling the POBox server with `sys:trace(PidOrName, true)` to see what transforms come out and give the corruption.

edescourtis commented:
Output of
```
        dbg:tracer(),
        dbg:p(Box, [timestamp, m, c]),
```
```
===> Retrying prop_pobox:prop_we_always_go_to_passive_mode_after_an_automatic_transfer()
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460218})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460273})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460293})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460305})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460315})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460322})
(<0.393.0>) << {'$gen_cast',{post,{msg,<<>>}}} (Timestamp: {1543,960956,
                                                            460340})
(<0.393.0>) << {'$gen_call',{<0.403.0>,#Ref<0.2286255986.979369985.178366>},
                            {resize,5}} (Timestamp: {1543,960956,460353})
(<0.393.0>) << {'$gen_call',{<0.404.0>,#Ref<0.2286255986.979369985.178367>},
                            {resize,1}} (Timestamp: {1543,960956,460361})
(<0.393.0>) <0.392.0> ! {mail,<0.393.0>,new_data} (Timestamp: {1543,960956,
                                                               460380})
(<0.393.0>) <0.403.0> ! {#Ref<0.2286255986.979369985.178366>,ok} (Timestamp: {1543,
                                                                              960956,
                                                                              460423})
(<0.393.0>) error_logger ! {notify,
                            {error,<0.52.0>,
                             {<0.393.0>,
                              "** State machine ~tp terminating~n** Last event = ~tp~n** When server state  = ~tp~n** Reason for termination = ~w:~tp~n** Callback mode = ~p~n** Stacktrace =~n**  ~tp~n",
                              [<0.393.0>,
                               {{call,
                                 {<0.404.0>,
                                  #Ref<0.2286255986.979369985.178367>}},
                                {resize,1}},
                               {passive,
                                {state,
                                 {buf,keep_old,5,5,2,
                                  {[{msg,<<>>}],[{msg,<<>>}]}},
                                 <0.392.0>,undefined,undefined,<0.392.0>,
                                 #Ref<0.2286255986.979369985.178349>,<0.5.0>,
                                 undefined,<0.393.0>}},
                               error,badarg,state_functions,
                               [{queue,split,
                                 [4,{[{msg,<<>>}],[{msg,<<>>}]}],
                                 [{file,"queue.erl"},{line,325}]},
                                {pobox,drop,4,
                                 [{file,
                                   "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                  {line,499}]},
                                {pobox,resize_buf,2,
                                 [{file,
                                   "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                  {line,448}]},
                                {pobox,handle_call,4,
                                 [{file,
                                   "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                  {line,344}]},
                                {gen_statem,call_state_function,5,
                                 [{file,"gen_statem.erl"},{line,1633}]},
                                {gen_statem,loop_event_state_function,6,
                                 [{file,"gen_statem.erl"},{line,1023}]},
                                {proc_lib,init_p_do_apply,3,
                                 [{file,"proc_lib.erl"},{line,247}]}]]}}} (Timestamp: {1543,
                                                                                       960956,
                                                                                       460462})
(<0.393.0>) code_server ! {code_call,<0.393.0>,{ensure_loaded,sys}} (Timestamp: {1543,
                                                                                 960956,
                                                                                 460504})
(<0.393.0>) << {'DOWN',#Ref<0.2286255986.979369985.178349>,process,<0.392.0>,
                       normal} (Timestamp: {1543,960956,460559})
(<0.393.0>) << {code_server,{module,sys}} (Timestamp: {1543,960956,463028})
(<0.393.0>) error_logger ! {notify,
                            {error_report,<0.52.0>,
                             {<0.393.0>,crash_report,
                              [[{initial_call,{pobox,init,['Argument__1']}},
                                {pid,<0.393.0>},
                                {registered_name,[]},
                                {error_info,
                                 {error,badarg,
                                  [{queue,split,
                                    [4,{[{msg,<<>>}],[{msg,<<>>}]}],
                                    [{file,"queue.erl"},{line,325}]},
                                   {pobox,drop,4,
                                    [{file,
                                      "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                     {line,499}]},
                                   {pobox,resize_buf,2,
                                    [{file,
                                      "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                     {line,448}]},
                                   {pobox,handle_call,4,
                                    [{file,
                                      "/home/eric/dev/pobox/_build/test/lib/pobox/src/pobox.erl"},
                                     {line,344}]},
                                   {gen_statem,call_state_function,5,
                                    [{file,"gen_statem.erl"},{line,1633}]},
                                   {gen_statem,loop_event_state_function,6,
                                    [{file,"gen_statem.erl"},{line,1023}]},
                                   {proc_lib,init_p_do_apply,3,
                                    [{file,"proc_lib.erl"},{line,247}]}]}},
                                {ancestors,[<0.392.0>,<0.5.0>]},
                                {message_queue_len,1},
                                {messages,
                                 [{'DOWN',
                                   #Ref<0.2286255986.979369985.178349>,
                                   process,<0.392.0>,normal}]},
                                {links,[]},
                                {dictionary,[]},
                                {trap_exit,false},
                                {status,running},
                                {heap_size,1598},
                                {stack_size,27},
                                {reductions,657}],
                               []]}}} (Timestamp: {1543,960956,463085})


```

edescourtis commented:
```
eric@eric-desktop:~/dev/pobox$ rebar3 shell
===> Verifying dependencies...
===> Compiling pobox
Erlang/OTP 20 [erts-9.3.3.5] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:0] [hipe] [kernel-poll:false]

Eshell V9.3.3.5  (abort with ^G)
1> {ok, Box} = pobox:start_link(#{max => 10, type => keep_old}).
{ok,<0.120.0>}
2> pobox:post(Box, {msg, <<>>}).
ok
3> pobox:post(Box, {msg, <<>>}).
ok
4> pobox:post(Box, {msg, <<>>}).
ok
5> pobox:post(Box, {msg, <<>>}).
ok
6> pobox:post(Box, {msg, <<>>}).
ok
7> pobox:post(Box, {msg, <<>>}).
ok
8> pobox:post(Box, {msg, <<>>}).
ok
9> l(sys).
{module,sys}
10> sys:get_state(Box).
{passive,{state,{buf,keep_old,10,7,0,
                     {[{msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>}],
                      [{msg,<<>>}]}},
                <0.117.0>,undefined,undefined,<0.117.0>,undefined,undefined,
                undefined,<0.120.0>}}
11> pobox:resize(Box, 5), pobox:resize(Box, 1).

=ERROR REPORT==== 4-Dec-2018::17:11:44 ===
** State machine <0.120.0> terminating
** Last event = {{call,{<0.117.0>,#Ref<0.879056874.711983105.214502>}},
                 {resize,1}}
** When server state  = {passive,
                            {state,
                                {buf,keep_old,5,5,2,
                                    {[{msg,<<>>}],[{msg,<<>>}]}},
                                <0.117.0>,undefined,undefined,<0.117.0>,
                                undefined,undefined,undefined,<0.120.0>}}
** Reason for termination = error:badarg
** Callback mode = state_functions
** Stacktrace =
**  [{queue,split,
            [4,{[{msg,<<>>}],[{msg,<<>>}]}],
            [{file,"queue.erl"},{line,325}]},
     {pobox,drop,4,
            [{file,"/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl"},
             {line,499}]},
     {pobox,resize_buf,2,
            [{file,"/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl"},
             {line,448}]},
     {pobox,handle_call,4,
            [{file,"/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl"},
             {line,344}]},
     {gen_statem,call_state_function,5,[{file,"gen_statem.erl"},{line,1633}]},
     {gen_statem,loop_event_state_function,6,
                 [{file,"gen_statem.erl"},{line,1023}]},
     {proc_lib,init_p_do_apply,3,[{file,"proc_lib.erl"},{line,247}]}]
** exception exit: badarg
     in function  queue:split/2
        called as queue:split(4,{[{msg,<<>>}],[{msg,<<>>}]})
     in call from pobox:drop/4 (/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl, line 499)
     in call from pobox:resize_buf/2 (/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl, line 448)
     in call from pobox:handle_call/4 (/home/eric/dev/pobox/_build/default/lib/pobox/src/pobox.erl, line 344)
     in call from gen_statem:call_state_function/5 (gen_statem.erl, line 1633)
     in call from gen_statem:loop_event_state_function/6 (gen_statem.erl, line 1023)
     in call from proc_lib:init_p_do_apply/3 (proc_lib.erl, line 247)
12> 

```

edescourtis commented:
Very strange!:
```erlang
1> {ok, Box} = pobox:start_link(#{max => 10, type => keep_old}).
{ok,<0.115.0>}
2> [pobox:post(Box, {msg, <<>>}) || _ <- lists:seq(1, 7)].
[ok,ok,ok,ok,ok,ok,ok]
3> sys:get_state(Box).
{passive,{state,{buf,keep_old,10,7,0,
                     {[{msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>},
                       {msg,<<>>}],
                      [{msg,<<>>}]}},
                <0.112.0>,undefined,undefined,<0.112.0>,undefined,undefined,
                undefined,<0.115.0>}}
4> pobox:resize(Box, 2). 
ok
5> sys:get_state(Box).
{passive,{state,{buf,keep_old,2,2,5,
                     {[{msg,<<>>},{msg,<<>>},{msg,<<>>},{msg,<<>>}],
                      [{msg,<<>>}]}},
                <0.112.0>,undefined,undefined,<0.112.0>,undefined,undefined,
                undefined,<0.115.0>}}
6> pobox:resize(Box, 1).
ok
7> sys:get_state(Box).  
{passive,{state,{buf,keep_old,1,1,6,
                     {[{msg,<<>>},{msg,<<>>},{msg,<<>>}],[{msg,<<>>}]}},
                <0.112.0>,undefined,undefined,<0.112.0>,undefined,undefined,
                undefined,<0.115.0>}}

```

edescourtis commented:
Found it! It was:
```erlang
drop(keep_old, N, Size, Queue) ->
    if Size > N  -> element(1, queue:split(N, Queue));
       Size =< N -> queue:new()
    end;
```
instead of: 
```erlang
drop(keep_old, N, Size, Queue) ->
    if Size > N  -> element(1, queue:split(Size - N, Queue));
       Size =< N -> queue:new()
    end;
```

edescourtis committed:
Added property based tests

 * Fixed a drop when using keep_old buffer
 * Check that we always return to passive mode (for both give_away and heir)
 * Check that we reply 'full' when the PO Box has reached capacity
 * Fixed incorrect documentation in README.md (missing Pid in message)

edescourtis commented:
Alright, everything seems to be [passing](https://travis-ci.com/edescourtis/pobox) now.

edescourtis commented on src/pobox.erl:
```diff
@@ -491,7 +496,7 @@ drop(stack, N, Size, L) ->
     end;
 drop(keep_old, N, Size, Queue) ->
-    if Size > N  -> element(1, queue:split(N, Queue));
+    if Size > N  -> element(1, queue:split(Size - N, Queue));
```
This fixes the resize issue. I guess this explains why it was happening only in `keep_old`.

ferd replied:
Nice. That was a tricky rare-to-trigger bug.

edescourtis commented:
Can we mark this as done for now and tag the release? I need to get back to some other work. I will return to adding more tests at a later date. Plus I think I need to get a copy of your book and read it.

ferd commented:
Done and published on hex as well.

edescourtis commented:
Thanks! Don't forget the CI account when you get the chance to set that up.
